mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -375,6 +386,203 @@ pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
     }
 }
 
+/// Represents a control statement that can be used to break or continue a UTXO loop.
+pub enum UtxoLoopCtrlStatement {
+    /// Signals that the UTXO loop should be terminated immediately, with the specified error message.
+    Break(String),
+    /// Signals that the current iteration of the UTXO loop should be skipped over, with the specified error message.
+    Continue(String),
+}
+
+async fn from_block_height(
+    client: &ElectrumClient,
+    spv_conf: &SPVConf,
+    storage: &dyn BlockHeaderStorageOps,
+    ticker: &str,
+) -> Result<u64, UtxoLoopCtrlStatement> {
+    let height = match storage.get_last_block_height().await {
+        Ok(Some(height)) => height,
+        Ok(None) => {
+            if let Err(err) = validate_and_store_starting_header(client, ticker, storage, spv_conf).await {
+                return Err(UtxoLoopCtrlStatement::Break(err.to_string()));
+            }
+            spv_conf.starting_block_header.height
+        },
+        Err(err) => return Err(UtxoLoopCtrlStatement::Continue(err.to_string())),
+    };
+
+    Ok(height)
+}
+
+async fn to_block_height(
+    coin: &UtxoCoinFields,
+    from_block_height: u64,
+    chunk_size: u64,
+    block_count: &mut u64,
+    ticker: &str,
+) -> Result<u64, UtxoLoopCtrlStatement> {
+    let mut to_block_height = from_block_height + chunk_size;
+    if to_block_height > *block_count {
+        *block_count = match coin.rpc_client.get_block_count().compat().await {
+            Ok(h) => h,
+            Err(err) => {
+                error!("Error {err:} on getting the height of the latest {ticker} block from rpc!");
+                return Err(UtxoLoopCtrlStatement::Continue(err.to_string()));
+            },
+        };
+
+        // More than `chunk_size` blocks could have appeared since the last `get_block_count` RPC.
+        // So reset `to_block_height` if only `from_block_height + chunk_size > actual_block_count`.
+        if to_block_height > *block_count {
+            to_block_height = *block_count;
+        }
+    }
+
+    drop_mutability!(to_block_height);
+    Ok(to_block_height)
+}
+
+async fn retrieve_headers_helper(
```

Comment:
A retrieve headers loop can be done inside this function instead of returning `UtxoLoopCtrlStatement` error and continuing/breaking the outside loop. You will not need `UtxoLoopCtrlStatement` error after that since error returning from this and `retrieve_and_revalidate_mismatching_header` will be to break the outside loop only.

Reply:
done